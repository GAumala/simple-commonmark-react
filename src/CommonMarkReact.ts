import * as Commonmark from 'commonmark';
import { ReactElement } from 'react';

import ContainerStack from './ContainerStack';
import RendererSelector from './renderers/RendererSelector';
import RenderOptions from './RenderOptions';

const parser = new Commonmark.Parser();

const isValidReactElement = (
  element: string | ReactElement<any> | null,
): element is ReactElement<any> => {
  return element != null && (element as ReactElement<any>).props !== undefined;
};

const isValidContainer = (node: Commonmark.Node): boolean => {
  // while images are markdown containers, they are not react containers (can't have children)
  return node.isContainer && node.type !== 'image';
};

const insertNewElementToResult = (
  result: Array<ReactElement<any>>,
  stack: ContainerStack,
  newElement: string | ReactElement<any> | null,
) => {
  const currentContainer = stack.peekCurrentContainer();
  if (currentContainer && newElement !== null)
    currentContainer.props.children.push(newElement);
  else if (isValidReactElement(newElement)) result.push(newElement);
};

const updateContainerStack = (
  containerStack: ContainerStack,
  node: Commonmark.Node,
  newElement: string | ReactElement<any> | null,
) => {
  if (isValidContainer(node) && isValidReactElement(newElement))
    containerStack.pushContainer(node.type, newElement);
};

const computeUniqueKey = (
  result: Array<ReactElement<any>>,
  containerStack: ContainerStack,
): string => {
  const currentContainer = containerStack.peekCurrentContainer();
  if (currentContainer) return '' + currentContainer.props.children.length;
  return '' + result.length;
};

export const renderNodes = (
  source: string,
  options: RenderOptions | undefined,
): Array<ReactElement<any>> => {
  const result: Array<ReactElement<any>> = [];
  const nodesBlock = parser.parse(source);
  const walker = nodesBlock.walker();
  const containerStack = new ContainerStack();
  let event: Commonmark.NodeWalkingStep = walker.next();

  while (event) {
    const currentNode = event.node;
    if (event.entering) {
      const key = computeUniqueKey(result, containerStack);
      const renderer = RendererSelector(currentNode, options);
      const currentReactElement = renderer.renderNode(key);
      insertNewElementToResult(result, containerStack, currentReactElement);
      updateContainerStack(containerStack, currentNode, currentReactElement);
    } else if (currentNode.isContainer)
      containerStack.removeContainer(currentNode.type);
    event = walker.next();
  }
  return result;
};
