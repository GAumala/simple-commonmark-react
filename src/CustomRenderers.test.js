import React from 'react';
import renderer from 'react-test-renderer';

import { renderNodes } from './CommonMarkReact.ts';

const renderToJSON = (markdown, rendererKey, customRenderer) => {
  const options = { customRenderers: {} };
  options.customRenderers[rendererKey] = customRenderer;
  const nodes = renderNodes(markdown, options);
  const rootElement = React.createElement('div', null, nodes);
  const component = renderer.create(rootElement);
  return component.toJSON();
};

function GenericRenderer(className) {
  return function() {
    return {
      renderNode: function(key) {
        return React.createElement('div', { key, className }, []);
      },
    };
  };
}

test('Renders custom bold', () => {
  const customRenderer = GenericRenderer('bold');
  const markdown = '**hello world**\n';
  const tree = renderToJSON(markdown, 'strong', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom code block', () => {
  const customRenderer = GenericRenderer('code-block');
  const markdown = '```\nhello world\n```';
  const tree = renderToJSON(markdown, 'code_block', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom inline code', () => {
  const customRenderer = GenericRenderer('code');
  const markdown = '`hello world`';
  const tree = renderToJSON(markdown, 'code', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom header', () => {
  const customRenderer = GenericRenderer('header');
  const markdown = '# hello world';
  const tree = renderToJSON(markdown, 'heading', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom image', () => {
  const customRenderer = GenericRenderer('image');
  const markdown = '![hello world](/pic.png)';
  const tree = renderToJSON(markdown, 'image', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom italics', () => {
  const customRenderer = GenericRenderer('italics');
  const markdown = '*hello world*\n';
  const tree = renderToJSON(markdown, 'emph', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom linebreak', () => {
  const customRenderer = GenericRenderer('line-break');
  const markdown = 'hello\nworld';
  const tree = renderToJSON(markdown, 'softbreak', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom link', () => {
  const customRenderer = GenericRenderer('link');
  const markdown = '[hello world](/home)';
  const tree = renderToJSON(markdown, 'link', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom list', () => {
  const customRenderer = GenericRenderer('list');
  const markdown = '- hello world\n';
  const tree = renderToJSON(markdown, 'list', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom list item', () => {
  const customRenderer = GenericRenderer('list-item');
  const markdown = '- hello world\n';
  const tree = renderToJSON(markdown, 'item', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom paragraph', () => {
  const customRenderer = GenericRenderer('paragraph');
  const markdown = 'hello world\n';
  const tree = renderToJSON(markdown, 'paragraph', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom block quote', () => {
  const customRenderer = GenericRenderer('block_quote');
  const markdown = '> hello world\n';
  const tree = renderToJSON(markdown, 'block_quote', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom text', () => {
  const customRenderer = GenericRenderer('text');
  const markdown = 'hello world\n';
  const tree = renderToJSON(markdown, 'text', customRenderer);
  expect(tree).toMatchSnapshot();
});

test('Renders custom thematic break', () => {
  const customRenderer = GenericRenderer('thematic_break');
  const markdown = '***';
  const tree = renderToJSON(markdown, 'thematic_break', customRenderer);
  expect(tree).toMatchSnapshot();
});
