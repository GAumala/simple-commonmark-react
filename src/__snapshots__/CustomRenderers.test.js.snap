// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders custom block quote 1`] = `
<div>
  <div
    className="block_quote"
  >
    <p
      className={undefined}
    >
      hello world
    </p>
  </div>
</div>
`;

exports[`Renders custom bold 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="bold"
    >
      hello world
    </div>
  </p>
</div>
`;

exports[`Renders custom code block 1`] = `
<div>
  <div
    className="code-block"
  />
</div>
`;

exports[`Renders custom header 1`] = `
<div>
  <div
    className="header"
  >
    hello world
  </div>
</div>
`;

exports[`Renders custom image 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="image"
    />
    hello world
  </p>
</div>
`;

exports[`Renders custom inline code 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="code"
    />
  </p>
</div>
`;

exports[`Renders custom italics 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="italics"
    >
      hello world
    </div>
  </p>
</div>
`;

exports[`Renders custom linebreak 1`] = `
<div>
  <p
    className={undefined}
  >
    hello
    <div
      className="line-break"
    />
    world
  </p>
</div>
`;

exports[`Renders custom link 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="link"
    >
      hello world
    </div>
  </p>
</div>
`;

exports[`Renders custom list 1`] = `
<div>
  <div
    className="list"
  >
    <li
      className={undefined}
    >
      hello world
    </li>
  </div>
</div>
`;

exports[`Renders custom list item 1`] = `
<div>
  <ul
    className={undefined}
  >
    <div
      className="list-item"
    >
      hello world
    </div>
  </ul>
</div>
`;

exports[`Renders custom paragraph 1`] = `
<div>
  <div
    className="paragraph"
  >
    hello world
  </div>
</div>
`;

exports[`Renders custom text 1`] = `
<div>
  <p
    className={undefined}
  >
    <div
      className="text"
    />
  </p>
</div>
`;
